src/beast/test/asio/streambuf.cpp

```diff
@@ -0,0 +1,296 @@
+            sb.commit(buffer_copy(d, buffer(s.data()+x+y, z)));
+            sb.commit(2);
+            expect(sb.size() == x + y + z);
+            expect(buffer_size(sb.data()) == sb.size());
+            expect(to_string(sb.data()) == s);
+            sb.consume(t);
+            d = sb.prepare(0); expect(buffer_size(d) == 0);
+            expect(to_string(sb.data()) == s.substr(t, std::string::npos));
+            sb.consume(u);
+            expect(to_string(sb.data()) == s.substr(t + u, std::string::npos));
+            sb.consume(v);
+            expect(to_string(sb.data()) == "");
+            sb.consume(1);
+            d = sb.prepare(0); expect(buffer_size(d) == 0);
+        }
+        }}}}}
+    }
+
+    template<class Alloc1, class Alloc2>
+    static
+    bool
+    eq(basic_streambuf<Alloc1> const& sb1,
+        basic_streambuf<Alloc2> const& sb2)
+    {
+        return to_string(sb1.data()) == to_string(sb2.data());
+    }
+
+    void testSpecial()
+    {
+        using boost::asio::buffer;
+        using boost::asio::buffer_cast;
+        using boost::asio::buffer_size;
+        std::string const s = "Hello, world";
+        expect(s.size() == 12);
+        for(std::size_t i = 1; i < 12; ++i) {
+        for(std::size_t x = 1; x < 4; ++x) {
+        for(std::size_t y = 1; y < 4; ++y) {
+        std::size_t z = s.size() - (x + y);
+        {
+            streambuf sb(i);
+            sb.commit(buffer_copy(sb.prepare(x), buffer(s.data(), x)));
+            sb.commit(buffer_copy(sb.prepare(y), buffer(s.data()+x, y)));
+            sb.commit(buffer_copy(sb.prepare(z), buffer(s.data()+x+y, z)));
+            expect(to_string(sb.data()) == s);
+            {
+                streambuf sb2(sb);
+                expect(eq(sb, sb2));
+            }
+            {
+                streambuf sb2;
+                sb2 = sb;
+                expect(eq(sb, sb2));
+            }
+            {
+                streambuf sb2(std::move(sb));
+                expect(to_string(sb2.data()) == s);
+                expect(buffer_size(sb.data()) == 0);
+                sb = std::move(sb2);
+                expect(to_string(sb.data()) == s);
+                expect(buffer_size(sb2.data()) == 0);
+            }
+        }
+        }}}
+    }
+
+    void testAllocator()
+    {
+        {
+            using alloc_type =
+                test_allocator<char, false, false, false, false>;
+            using sb_type = basic_streambuf<alloc_type>;
+            sb_type sb;
+            expect(sb.get_allocator().id() == 1);
+        }
+        {
+            using alloc_type =
+                test_allocator<char, false, false, false, false>;
+            using sb_type = basic_streambuf<alloc_type>;
+            sb_type sb;
+            expect(sb.get_allocator().id() == 2);
+            sb_type sb2(sb);
+            expect(sb2.get_allocator().id() == 2);
+            sb_type sb3(sb, alloc_type{});
+            //expect(sb3.get_allocator().id() == 3);
```

Comment:
I haven't tested this out, but what about putting `id_` in a base class under `test_allocator_base` which is not templated on anything at all (not even `T`)?